server/watchdog/watchdog.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchdog
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+const (
+	tickMs = 100
+)
+
+var (
+	wdInstance     *watchdog
+	nextActivityId uint64
+)
+
+type activity struct {
+	id              uint64
+	name            string
+	inactiveElapsed int
+}
+
+type watchdog struct {
```

Comment:
Watchdog should have an abstracted (golang) interface in a separate file... 
and potentially multiple implementations.